hydro_lang/src/ir.rs

```diff
@@ -2277,6 +2301,113 @@ impl HydroNode {
                 (fold_ident, input_location_id)
             }
 
+            HydroNode::ReduceKeyedWatermark {
+                f,
+                input,
+                watermark,
+                ..
+            } => {
+                let (input, lifetime) =
+                    if let HydroNode::Persist { inner: input, .. } = input.as_mut() {
+                        (input, quote!('static))
+                    } else {
+                        (input, quote!('tick))
+                    };
+
+                let (input_ident, input_location_id) =
+                    input.emit_core(builders_or_callback, built_tees, next_stmt_id);
+
+                let input_type = *input.metadata().output_type.clone().unwrap().0;
```

Comment:
This will be enforced by the Rust type signature so we don't _really_ need this.